src/lib.rs

```diff
@@ -137,23 +148,36 @@ pub fn formatter(
         configuration.language = l
     }
 
-    // All the work related to tree-sitter and the query is done here
-    log::info!("Apply Tree-sitter query");
-    let mut atoms = tree_sitter::apply_query(&content, &query, configuration.language)?;
+    match operation {
+        Operation::Format { skip_idempotence } => {
+            // All the work related to tree-sitter and the query is done here
+            log::info!("Apply Tree-sitter query");
+            let mut atoms = tree_sitter::apply_query(&content, &query, configuration.language)?;
 
-    // Various post-processing of whitespace
-    atoms.post_process();
+            // Various post-processing of whitespace
+            atoms.post_process();
 
-    // Pretty-print atoms
-    log::info!("Pretty-print output");
-    let rendered = pretty::render(&atoms[..], configuration.indent_level)?;
-    let trimmed = trim_whitespace(&rendered);
+            // Pretty-print atoms
+            log::info!("Pretty-print output");
+            let rendered = pretty::render(&atoms[..], configuration.indent_level)?;
+            let trimmed = trim_whitespace(&rendered);
 
-    if !skip_idempotence {
-        idempotence_check(&trimmed, &query, language)?
-    }
+            if !skip_idempotence {
+                idempotence_check(&trimmed, &query, language)?
+            }
+
+            write!(output, "{trimmed}")?;
+        }
 
-    write!(output, "{trimmed}")?;
+        Operation::Visualise { output_format } => {
+            let (tree, _) = tree_sitter::parse(&content, &configuration.language.grammars())?;
```

Comment:
`parse` should now just take the language (not the grammars) as input, so the API is similar to `apply_query` above.

Reply:
`apply_query` calls `parse` internally, which takes the grammars: 

https://github.com/tweag/topiary/blob/5e83b4612629de19ab7304dee9d15a584812a300/src/tree_sitter.rs#L81-L82

This was done for the `.ml`/`.mli` split. I'll refactor `parse` so it just takes the language, cycling through the grammars internally...